src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -3481,6 +3519,218 @@ QIcon ViewProviderSketch::mergeColorfulOverlayIcons (const QIcon & orig) const
     return Gui::ViewProvider::mergeColorfulOverlayIcons (mergedicon);
 }
 
+bool ViewProviderSketch::getCameraMaxDimension(const Gui::View3DInventorViewer* viewer) {
+    float fHeight = -1.0;
+    float fWidth = -1.0;
+    viewer->getDimensions(fHeight, fWidth);
+    if (fHeight < 0 || fWidth < 0)
+        return false;
+
+    fHeight = std::max(fHeight, fWidth);
+    if (fabs(fHeight - camMaxDimension) > 0) { //ie if user zoomed.
+        camMaxDimension = fHeight;
+        return true;
+    }
+    return false;
+}
+
+bool ViewProviderSketch::getCameraCenter(const Gui::View3DInventorViewer* viewer)
+{
+    SoCamera* cam = viewer->getSoRenderManager()->getCamera();
+    if(!cam)
+        return false;
+
+    SbVec3f direction;
+    cam->orientation.getValue().multVec(SbVec3f(0, 0, -1), direction);
+    SbVec3f newCamCenterOnSketch = cam->position.getValue() + cam->focalDistance.getValue() * direction;
+
+    if ((camCenterOnSketch - newCamCenterOnSketch).length() > 0.1 * camMaxDimension) { //ie if user moved by more than 10% of the current 3d view dimensions.
+        camCenterOnSketch = newCamCenterOnSketch;
+        return true;
+    }
+    return false;
+}
+
+double ViewProviderSketch::getGridSize(const Gui::View3DInventorViewer* viewer)
+{
+    short pixelWidth = -1;
+    short pixelHeight = -1;
+    viewer->getViewportRegion().getViewportSizePixels().getValue(pixelWidth, pixelHeight);
+    if (pixelWidth < 0 || pixelHeight < 0)
+        return GridSize.getValue();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    int numberOfLines = std::max(pixelWidth, pixelHeight) / hGrp->GetInt("GridSizePixelThreshold", 15);
+
+    double newGridSize = pow(hGrp->GetInt("GridNumberSubdivision", 10), 1 + floor(log10(camMaxDimension / numberOfLines)));
+
+    //cap the grid size
+    newGridSize = std::max(newGridSize, 0.00000001);

```

Comment:
this might need to be higher or dependent on the sketch scale and machine accuracy given this bug I encountered https://github.com/FreeCAD/FreeCAD/pull/7754#issuecomment-1316875634